### PR TITLE
CHAD-16364 & CHAD-16365: Add lockUsers and lockCredentials capability

### PR DESCRIPTION
**Pull request by pegor-karoglanian**
Check all that apply

# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [ ] I have reviewed the README.md file
          - [ ] I have reviewed the CODE_OF_CONDUCT.md file
          - [ ] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [ ] New feature
- [ ] Refactor

# Checklist

- [ ] I have performed a self-review of my code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have verified my changes by testing with a device or have communicated a plan for testing
- [ ] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change

# Summary of Completed Tests

### CODE REVIEW

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 1]
```diff
@@ -291,6 +293,35 @@ local name_slot = function(driver, device, command)
   end
 end
 
+local migrate = function(driver, device, command)
+  local lock_users = {}
+  local lock_credentails = {}
+  local index = 0
+  for code_slot, code_name in pairs(lock_utils.get_lock_codes(device)) do
+    table.insert(lock_users, {userIndex = index, userType = "guest", userName = code_name})
+    table.insert(lock_credentails, {userIndex = index, credentialIndex = code_slot, crednetialType = "pin"})
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/init.lua]:
```suggestion
    table.insert(lock_credentails, {userIndex = index, credentialIndex = code_slot, crednentialType = "pin"})
```

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 2]
```diff
@@ -291,6 +293,35 @@ local name_slot = function(driver, device, command)
...
+    index = index + 1
+  end
+
+  local code_length  = device:get_field(capabilities.lockCodes.codeLength)
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: here you would want to be pulling from the device state store with 

`device:get_latest_state("main", lockCodes.ID, lockCodes.codeLength.NAME)`

`get_latest_state` gives you the most recent value of a capability attribute for a given device

File: drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_slga_migration.lua [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        mock_datastore.__assert_device_store_contains(mock_device.id, "migrationComplete", true)
+
+        -- Set min/max code length attributes
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.MinPINCodeLength:build_test_attr_report(mock_device, 5) })
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.MaxPINCodeLength:build_test_attr_report(mock_device, 10) })
+        test.socket.zigbee:__queue_receive({ mock_device.id, DoorLock.attributes.NumberOfPINUsersSupported:build_test_attr_report(mock_device, 4) })
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.minCodeLength(5,  { visibility = { displayed = false } })))
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.maxCodeLength(10,  { visibility = { displayed = false } })))
+        test.socket.capability:__expect_send( mock_device:generate_test_message("main", capabilities.lockCodes.maxCodes(4,  { visibility = { displayed = false } })))
+
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_slga_migration.lua]: I think you want a `test.wait_for_events()` here. I was getting occasional failures without it.

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 4]
```diff
@@ -291,6 +293,42 @@ local name_slot = function(driver, device, command)
...
+  local lock_credentials = {}
+  local lock_codes = lock_utils.get_lock_codes(device)
+  local ordered_codes = {}
+
+  for code in pairs(lock_codes) do
+      table.insert(ordered_codes, code)
+  end
+
+  table.sort(ordered_codes)
+  for index = 1, #ordered_codes do
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: I don't think this will work on a real device. The code indices aren't guaranteed to be contiguous.

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 5]
```diff
@@ -291,6 +293,42 @@ local name_slot = function(driver, device, command)
...
 end
 
+local migrate = function(driver, device, command)
+  local lock_users = {}
+  local lock_credentials = {}
+  local lock_codes = lock_utils.get_lock_codes(device)
+  local ordered_codes = {}
+
+  for code in pairs(lock_codes) do
+      table.insert(ordered_codes, code)
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: nit: you've got some inconsistencies with your indents

we usually use 2 spaces, for reference

File: drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_slga_migration.lua [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+local mock_datastore = require "integration_test.mock_env_datastore"
+
+local mock_device = test.mock_device.build_test_zigbee_device(
+    {
+      profile = t_utils.get_profile_definition("base-lock.yml"),
+      data = {
+        lockCodes = json.encode({
+          ["1"] = "Zach",
+          ["2"] = "Steven"
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_slga_migration.lua]: would it make sense to have non-contiguous codes here so that the user indices don't match the credential indices, just as a more corner-case-y test?

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 7]
```diff
@@ -291,6 +293,41 @@ local name_slot = function(driver, device, command)
...
+  local lock_codes = lock_utils.get_lock_codes(device)
+  local ordered_codes = {}
+
+  for code in pairs(lock_codes) do
+    table.insert(ordered_codes, code)
+  end
+
+  table.sort(ordered_codes)
+  for index, code_slot in ipairs(ordered_codes) do
+    table.insert(lock_users, {userIndex = index, userType = "guest", userName = lock_codes[code_slot]})
```
Comment (tpmanley) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: Zigbee supports other user types, and IIRC some locks like Yale do use them. It doesn't look like `lockCodes` kept track of that information, but could we try to get the type with the `Get User Type` command?

File: drivers/SmartThings/zwave-lock/src/init.lua [hunk 8]
```diff
@@ -146,6 +146,55 @@ local function update_codes(driver, device, cmd)
...
+  local lock_credentials = {}
+  local lc_data = json.decode(device.data.lockCodes)
+  local lock_codes = {}
+  local ordered_codes = {}
+  for k, v in pairs(lc_data) do
+    lock_codes[k] = v
+  end
+
+  for code in pairs(lock_codes) do
+      table.insert(ordered_codes, code)
```
Comment (greens) [inline drivers/SmartThings/zwave-lock/src/init.lua]: nit: spacing again

File: drivers/SmartThings/zwave-lock/src/init.lua [hunk 9]
```diff
@@ -146,6 +146,55 @@ local function update_codes(driver, device, cmd)
...
+  table.sort(ordered_codes)
+  for index = 1, #ordered_codes do
+      local code_slot, code_name = ordered_codes[index], lock_codes[ ordered_codes[index] ]
+      table.insert(lock_users, {userIndex = index, userType = "guest", userName = code_name})
+      table.insert(lock_credentials, {userIndex = index, credentialIndex = tonumber(code_slot), credentialType = "pin"})
+  end
+
+  local code_length  = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.codeLength.NAME)
+  local min_code_len = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.minCodeLength.NAME)
+  local max_code_len = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.maxCodeLength.NAME)
```
Comment (greens) [inline drivers/SmartThings/zwave-lock/src/init.lua]: the 4th (optional) argument to `get_latest_state` is a default value. 

you could save yourself a few lines

File: drivers/SmartThings/zigbee-lock/src/init.lua (hunk 7 above)
Comment (greens) [reply drivers/SmartThings/zigbee-lock/src/init.lua]: My gut says that we should keep it simple for the migration (i.e. limit device communication), but perhaps handle it in the new capability handlers.

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 10]
```diff
@@ -291,6 +293,41 @@ local name_slot = function(driver, device, command)
...
+
+  local code_length  = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.codeLength.NAME)
+  local max_code_len = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.maxCodeLength.NAME)
+  local min_code_len = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.minCodeLength.NAME)
+  local max_codes    = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.maxCodes.NAME)
+
+  if (code_length ~= nil) then
+    max_code_len = code_length
+    min_code_len = code_length
+  end
```
Comment (wkhenon) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: In the event that `code_length == nil`, would it make sense to set the `min_code_len` to 4 and the `max_code_len` to 8 (per the spec; how it's being done with Z-Wave)?

zcl-8-door-lock

Comment (pegor-karoglanian) [reply drivers/SmartThings/zigbee-lock/src/init.lua]: Good call, I'll set the min/max to 4/8 if they are unset.

File: drivers/SmartThings/zwave-lock/src/init.lua (hunk 9 above)
Comment (pegor-karoglanian) [reply drivers/SmartThings/zwave-lock/src/init.lua]: oo nice, thanks for headsup

File: drivers/SmartThings/zigbee-lock/src/init.lua [hunk 11]
```diff
@@ -291,6 +293,40 @@ local name_slot = function(driver, device, command)
...
+
+  table.sort(ordered_codes)
+  for index, code_slot in ipairs(ordered_codes) do
+    table.insert(lock_users, {userIndex = index, userType = "guest", userName = lock_codes[code_slot]})
+    table.insert(lock_credentials, {userIndex = index, credentialIndex = tonumber(code_slot), credentialType = "pin"})
+  end
+
+  local code_length  = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.codeLength.NAME)
+  local max_code_len = device:get_latest_state("main", capabilities.lockCodes.ID, capabilities.lockCodes.maxCodeLength.NAME, 4)
```
Comment (tpmanley) [inline drivers/SmartThings/zigbee-lock/src/init.lua]: Are the min/max code length defaults swapped?

Comment (pegor-karoglanian) [reply drivers/SmartThings/zigbee-lock/src/init.lua]: Whoops, good catch. I'll fix this.

**Comment by pegor-karoglanian** [comment]
12/10 update: Converted this PR to draft since a lot more is now going in. The latest commit adds all the commands for `lockUsers` and `lockCredentials` along with unit tests covering most of them. There are still more test cases I need to add and to tweak some of the functionality.

File: drivers/SmartThings/zigbee-lock/src/using-old-capabilities/init.lua
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/using-old-capabilities/init.lua:1]: so the old capability needs to continue to work _until_ `migrate()` is called, so that behavior has to remain

File: drivers/SmartThings/zigbee-lock/src/using-new-capabilities/samsungsds/init.lua [hunk 12]
```diff
@@ -0,0 +1,121 @@
+-- Copyright 2022 SmartThings
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/using-new-capabilities/samsungsds/init.lua]: nit, but this date should be updated

File: drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+  if max_users == 0 or available_index == nil then
+    -- Can't add any users - update commandResult statusCode
+    status_code = new_lock_utils.STATUS_RESOURCE_EXHAUSTED
+  else
+    -- use the passed in index if it's set
+    if user_index ~= nil then
+      available_index = user_index
+    end
+    table.insert(current_users, {userIndex = available_index, userType = user_type, userName = user_name})
+    device:set_field(new_lock_utils.LOCK_USERS, current_users)
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua]: for lock users, you probably want to use `persist = true` so it persists over hub restarts, since we cannot glean this info from the lock itself (or use the capability data store if that persists)

File: drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+  return new_lock_utils.STATUS_SUCCESS
+end
+
+new_lock_utils.delete_credential = function(device, credential_index) 
+  local credentials = new_lock_utils.get_credentials(device)
+  local status_code = new_lock_utils.STATUS_FAILURE
+
+  for index, credential in ipairs(credentials) do
+      if credential.userIndex == credential_index then
+          table.remove(credentials, index)
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua]: nit: whitespace

File: drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+local utils = require "st.utils"
+local capabilities = require "st.capabilities"
+local json = require "st.json"
+local LockCredentials = capabilities.lockCredentials
+local LockUsers = capabilities.lockUsers
+local INITIAL_INDEX = 1
+
+local new_lock_utils =  {
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/new_lock_utils.lua]: it's unlikely you'll need all of these strings, and, in general, they _should_ be available from inside the capabilities themselves.

File: drivers/SmartThings/zigbee-lock/src/using-new-capabilities/init.lua [hunk 16]
```diff
@@ -0,0 +1,444 @@
...
+
+    local credentials = lock_utils.get_credentials(device)
+    device:emit_event(capabilities.lockCredentials.credentials(credentials, {visibility = {displayed = false}}))
+  else
+    if lock_utils.get_credential(device, credential_index) ~= nil then
+      -- Credential has been deleted.
+      lock_utils.delete_credential(device, credential_index)
+      device:emit_event(capabilities.lockCredentials.credentials(lock_utils.get_credentials(device), {visibility = {displayed = false}}))
+    else 
+      -- unset?
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/using-new-capabilities/init.lua]: this would be a status `FAILURE` I believe

File: drivers/SmartThings/zigbee-lock/src/using-new-capabilities/init.lua [hunk 17]
```diff
@@ -0,0 +1,444 @@
...
+
+  if (zb_mess.body.zcl_body.user_status.value == UserStatusEnum.OCCUPIED_ENABLED) then   
+    if pending_credential ~= nil then
+      -- create credential
+      lock_utils.add_credential(device, pending_credential.userIndex,
+                                pending_credential.userType,
+                                pending_credential.credentialType, 
+                                credential_index)
+    else
+      -- update credential
```
Comment (greens) [inline drivers/SmartThings/zigbee-lock/src/using-new-capabilities/init.lua]: you probably want different handling based on whether `command` is an `add` or an `update`

File: drivers/SmartThings/zigbee-lock/src/using-old-capabilities/samsungsds/init.lua [hunk 18]
```diff
@@ -113,7 +113,10 @@ local samsung_sds_driver = {
     init = device_init
   },
   can_handle = function(opts, driver, device, ...)
-    return device:get_manufacturer() == "SAMSUNG SDS"
+    if device:get_manufacturer() == "SAMSUNG SDS" then
+      return true
+    end
+    return false
```
Comment (pegor-karoglanian) [inline drivers/SmartThings/zigbee-lock/src/using-old-capabilities/samsungsds/init.lua]: This is left over from some testing. Will clean up.

File: drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_migration.lua [hunk 19]
```diff
@@ -34,14 +34,16 @@ local mock_device = test.mock_device.build_test_zigbee_device(
           ["1"] = "Zach",
           ["2"] = "Steven"
         })
-      }
+      },
+      useOldCapabilityForTesting = true
```
Comment (pegor-karoglanian) [inline drivers/SmartThings/zigbee-lock/src/test/test_zigbee_lock_code_migration.lua:38]: Not sure if this is an appropriate way to go about this. Open to suggestions.
